### PR TITLE
Added toPaged, toBatched functions.

### PR DESCRIPTION
**Pull request by vkamiansky**
part of #8 
The function `toPaged` has been implemented by using recursion. The standard library already has a function `chunkBySize` that does the same. What is useful, though, about this implementation is that a simple tweak transforms it into `toBatched`.

### CODE REVIEW

**Comment by vkamiansky** [comment]
More testing for corner cases could be needed here. Especially, ToBatched batched is suspicious for adding more items to pages than allowed.

File: Composite/Composite/Transforms.fs [hunk 1]
```diff
@@ -35,10 +35,44 @@ module Transforms =
...
+                    | p, r -> yield p
+                              yield! getPages r
+            }
+        getPages inSeq
+
+    let toBatched batchSize getElemSize inSeq =
+        let rec getBatchAndRest numRemains pageAndRest = 
+            match numRemains > 0, pageAndRest with
+            | true, (p, r) -> match r |> Seq.tryHead with
+                                | Some h -> getBatchAndRest (numRemains-(getElemSize h)) (Array.append p [|h|], r |> Seq.tail)
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: A little style point. The parameter 'p' may be passed to Array.append by the pipe-forward operator while appending a new element to a batch.
The method toPaged may be changed in the same way.
Also, the value pageAndRest may be renamed to batchAndRest as it does in #15.

**Review by v-ilin — CHANGES_REQUESTED** [review]
Some recommendation may be applied.

File: Composite/Composite.Cs.Tests/Tests.cs [hunk 2]
```diff
@@ -15,6 +15,36 @@ internal class Simple {
...
+                .Take(2)
+                .ToArray();
+
+            Assert.Equal(1, batches[0][0].Number);
+            Assert.Equal(2, batches[0][1].Number);
+            Assert.Equal(3, batches[0][2].Number);
+            Assert.Equal(4, batches[0][3].Number);
+            Assert.Equal(5, batches[1][0].Number);
+            Assert.Equal(6, batches[1][1].Number);
+        }
```
Comment (v-ilin) [inline Composite/Composite.Cs.Tests/Tests.cs]: Looks like ToBatchedTest shows that the method AsBatched does not follow the described behavior because in the second batch the weight of elements greater than the batch size. Maybe we need to correct the method toBatched and add a check that current element size not greater than numRemains?

File: Composite/Composite/Transforms.fs [hunk 3]
```diff
@@ -35,10 +35,44 @@ module Transforms =
                     while cur |> Option.isSome do
                         yield cur |> (function | Some x -> x | None -> failwith "Unexpected situation in enumeration.")
                         cur <- getNext ()
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: Returning a current element to an output sequence may be done by taking a value of the option: `cur.Value`

Comment (vkamiansky) [reply Composite/Composite/Transforms.fs]: @v-ilin Makes sense. In either case, the only possible situation here has `cur.Value` valid at all times.

File: Composite/Composite/Transforms.fs [hunk 4]
```diff
@@ -57,12 +57,25 @@ module Transforms =
...
-                                | Some h -> getBatchAndRest (numRemains-(getElemSize h)) (Array.append p [|h|], r |> Seq.tail)
-                                | None -> pageAndRest
-            | false, pr -> pr
+        // Here we extract a batch from an input sequence.
+        // We do this recursively, passing the amount of remaining
+        // space in batch as a parameter, plus a tuple comprising
+        // the batch under constuction and the remaining sequence
+        // we have to walk through.
+        let rec getBatchAndRest roomLeft batchAndRest = 
+            let (page, rest) = batchAndRest
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: The value page may be renamed to batch

File: Composite/Composite/Transforms.fs [hunk 5]
```diff
@@ -57,12 +57,25 @@ module Transforms =
...
+            |> function
+               | Some head -> let size = getElemSize head
+                              let roomLeftNew = roomLeft - size
+                              page
+                              |> function
+                                 | [||] -> getBatchAndRest roomLeftNew ([|head|], rest |> Seq.tail)
+                                 | _ -> if roomLeftNew < 0
+                                        then batchAndRest
+                                        else getBatchAndRest roomLeftNew ([|head|] |> Array.append page, rest |> Seq.tail)
+               | None -> batchAndRest
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: In other cases with match expression, a branch that specifies a condition of ending a recursion goes first so an order may be changed here.

File: Composite/Composite/Transforms.fs [hunk 6]
```diff
@@ -57,12 +57,25 @@ module Transforms =
...
+        let rec getBatchAndRest roomLeft batchAndRest = 
+            let (page, rest) = batchAndRest
+            rest
+            |> Seq.tryHead
+            |> function
+               | Some head -> let size = getElemSize head
+                              let roomLeftNew = roomLeft - size
+                              page
+                              |> function
+                                 | [||] -> getBatchAndRest roomLeftNew ([|head|], rest |> Seq.tail)
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: In the case when a weight of the first element in a batch greater than a batch size it will be added to a batch without any checks.

File: Composite/Composite/Transforms.fs [hunk 7]
```diff
@@ -33,12 +33,60 @@ module Transforms =
...
+        getPages inSeq
+
+    let toBatched batchSize getElemSize inSeq =
+        // Here we extract a batch from an input sequence.
+        // We do this recursively, passing the amount of remaining
+        // space in batch as a parameter, plus a tuple comprising
+        // the batch under constuction and the remaining sequence
+        // we have to walk through.
+        let rec getBatchAndRest roomLeft batchAndRest = 
+            let (page, rest) = batchAndRest
```
Comment (v-ilin) [inline Composite/Composite/Transforms.fs]: The value `page` may be renamed to `batch`.

**Review by v-ilin — CHANGES_REQUESTED** [review]

File: Composite/Composite/Transforms.fs (hunk 6 above)
Comment (vkamiansky) [reply Composite/Composite/Transforms.fs]: @v-ilin This is precisely my idea here. A batch must contain at least one element, doesn't matter how heavy. Otherwise, it would only circle around in a meaningless loop. A multiple element batch cannot be of size greater than the provided batch size.

File: Composite/Composite/Transforms.fs (hunk 7 above)
Comment (vkamiansky) [reply Composite/Composite/Transforms.fs]: @v-ilin A valid point.

**Review by v-ilin — APPROVED** [review]
